Jaydel/implement reviews

jidbujawe opened the pull request:
# This PR branch is based off the jaydel/updates-after-meeting branch and consist the following changes:

- Created ReviewTest and implemented the TDD approach for developing Reviews.
- Continued on to create Controllers, Requests and Views.
- Updated routes to include reviews
- Edited layout of index and show
- Implemented front end of Review CRUD
- Star animation is still yet to be implemented in the forms. I think JS will have to be used for it. Placed the stars in the form anyway but will soon figure out how to use it as a slider for the rating or at least make it interactive.
- Re-run review test and passed
- Paginated review results in movies.show

jidbujawe committed:
Completed Reviews TDD and configured layout appropriately

jidbujawe committed:
Updated extend layout to app

jidbujawe committed:
Fixed review testing errors

jidbujawe committed:
Run npm install and npm run dev

jidbujawe committed:
Edited layout to add reviews link into menu

jidbujawe committed:
Finished implementing reviews CRUD

jidbujawe committed:
Passed review testing

jidbujawe committed:
Paginated review results in movies.show and linked each review

The paginator refreshes the page when going to the next page. Found a solution online that uses ajax to avoid this, will soon fix when we get to the JS part of the bootcamp.

jidbujawe committed:
Update create.blade.php

Updated cancel link to redirect back to the specific movie instead of going to index of reviews

jidbujawe committed:
Limited rating input to 2 decimals using regex

https://stackoverflow.com/questions/38947126/laravel-double-validation-regex-not-working

vrodriguero commented on app/Http/Controllers/ReviewController.php:
```diff
@@ -0,0 +1,70 @@
+{
+    public function index()
+    {
+        $reviews = Review::with(['movie'])
```
For this one, it may be best to ensure that you only getting the necessary columns to reduce the amount the processing. For example, let's say you have 40000 reviews, for each of these you will fetch the movie with all the columns, but in here you actually only uses maybe 2 of the movies table columns. 

So maybe in the `with` we can have something like 
```
with('movie:id,title,poster')
```
?

vrodriguero commented on app/Http/Controllers/ReviewController.php:
```diff
@@ -0,0 +1,70 @@
+
+    public function show(Review $review)
+    {
+        $review->with([
+            'movie'
+        ])->get();
```
same as the comment above :D

vrodriguero commented on app/Http/Controllers/ReviewController.php:
```diff
@@ -0,0 +1,70 @@
+        Review::create($request->validated());
+
+        return redirect()->route('reviews.index')
+                ->with(['message' => 'Your review has been added.']);
```
Optional - There is another way to do this to :D [Redirecting With Flashed Session Data]( https://laravel.com/docs/8.x/redirects#redirecting-with-flashed-session-data)

Maybe this makes it a tiny bit more clear on the controller and view ;D.

vrodriguero commented on app/Http/Middleware/Admin.php:
```diff
@@ -22,6 +22,6 @@ public function handle(Request $request, Closure $next)
         }
 
-        abort(403, 'Access denied');
+        return redirect('/');
```
perhaps you could also have a message to be displayed? Like "Oops! It seems like you do not have the right permissions to do this." hehe

vrodriguero commented on package-lock.json:
```diff
@@ -1,15357 +1,6 @@
-            }
-        }
-    },
+    "lockfileVersion": 1,
```
Hmmm not really sure why you and Kieran are having these changes on this file, but unless you have installed extra packages, this should not be commited as this auto generates when you run npm install.

jbujawe-cybercert replied:
Ah, I'm not sure why this is happening Vick. What do you suggest we should do? @vrodriguero

vrodriguero reviewed (changes requested):
Okay! This review is finished! 

I have made some suggestions on this one too Jayd, but I will wait for you to push the changes on the other branch you based this one and I then you can pull the changes here ;D

jidbujawe committed:
Merge branch 'main' into jaydel/implement-reviews

jidbujawe committed:
Implemented requested changes

- Used a filter for with('movie:id,title,poster') to lessen processing time
- Followed Redirecting With Flashed Session Data
- Added a message to admin middleware redirect

vrodriguero committed:
wip:
- removed Feature folder for tests
- Updated namespaces
- format code

vrodriguero committed:
revert package-lock.json

vrodriguero reviewed (changes requested):
Hey Jayd, 

I have reverted the changes on the package lock file. 
I also made some changes to the folder structure for the tests and updated the namespaces. 

I want you to take a look a the tests. There are a couple of tests failing and I think this is a good exercise for you.

It is always good to keep checking that your changes did not break any of the existing tests. I believe they would be related to the changes on the middleware and the addition of the banner column. 

Once all the tests pass you can request a re-review ;D Oh! Also, I notice a test with the `$this->withoutExceptionHandling` and I always forget that too, but generally we do not want to leave tests with that line in the project :D 

That is all! Keep up with the great work

jidbujawe committed:
Passed testing and followed structure of tests

jidbujawe commented:
Hi @vrodriguero, I got the tests to pass and followed the structure of the testing. Please let me know if what I did is correct? :) I've requested for another review now, my apologies forgot about the tests again! I'll make sure to review my work before asking for a PR! Also, please let me know if any of my code format is wrong so that I can follow the correct convention :) Thank you for all your help Vick!

vrodriguero commented on tests/Http/Middleware/AdminReviewTest.php:
```diff
@@ -0,0 +1,130 @@
+use App\Models\Review;
+use Illuminate\Foundation\Testing\RefreshDatabase;
+
+class AdminReviewTest extends TestCase
```
Hey Jayd, I think that the name here could be changed to AuthUser as we are not really testing if the admin can do reviews. 😄

jbujawe-cybercert replied:
Hi @vrodriguero, sorry I checked all the tests and it passed, would you mind letting me know which ones or send me a screenshot?

vrodriguero replied:
There are 10 tests failing in mine for the celeb test, that is because the admin middleware was changed. So, when we do a change we have to look at other parts that could be affected. So any test that it might be using the middleware should be checked :D

jbujawe-cybercert replied:
So sorry! I did not think to edit the Celeb test at all since Kieran wrote the tests and I'm responsible for the Movies and Reviews! From now on, I'll remember to edit all tests! Thank you @vrodriguero!

vrodriguero commented on tests/Http/Controllers/ReviewTest.php:
```diff
@@ -0,0 +1,324 @@
+    use RefreshDatabase;
+
+    /** @test */
+    public function an_admin_can_add_a_review()
```
Same here, does it matter to be the admin? I think we should test if only the authenticated user can create, and only those who uploaded the review should be able to edit/delete.

jbujawe-cybercert replied:
I also tested the authenticated users in the test. And I thought it would be good for an admin to be able to edit all reviews, not just their own for moderation? However, I'll try to do this then @vrodriguero :D

vrodriguero replied:
I think that an admin might be able to delete a review, but I think that editing is a bit much don't you think? For example, as a user I left a review saying that I did not like the movie, then the admin comes and changes the review to say something completely different. I will be pretty unsatisfied right? However, I do think that an admin should be able to delete comments that are harmful or inappropriate right?

jbujawe-cybercert replied:
Okay will do that Vick!

vrodriguero reviewed (changes requested):
Hey Jayd, only a few suggestions with the review tests ;D

I think because you have changed the admin middleware, the existing tests for the middleware are falling, so it is always good to check if all of the tests pass 😄

jidbujawe committed:
Implemented requested changes

- Followed testing folder structure
- Tested specifically for auth users only
- Tested users cannot edit a review they did not create
- Tested for admin to be able to delete any reviews

jidbujawe committed:
Merge branch 'main' into jaydel/implement-reviews

# Conflicts:
#	routes/web.php

jidbujawe committed:
Updated tests and passed

vrodriguero committed:
wip:
- Moved tests from middleware over to controller

vrodriguero reviewed (changes requested):
Jayd all seems okay!

I have made some changes to the structure of the files, as I have added a simple admin middleware test on Kieran's branch. So I think that it would be easier for you guys to just have one test file for each controller ;D For this reason, I moved the tests for movie and review from the middleware to the file on the controller. 

My suggestion would be to add some tests regarding what would happen if an auth user tries to delete or edit a review that belongs to another user. So I think we need to have a way to validate that 😄 

Sorry for all the suggestions hehhe

vrodriguero committed:
changes to resolve conflict

vrodriguero committed:
Merge branch 'main' into jaydel/implement-reviews

# Conflicts:
#	tests/Http/Middleware/AdminMiddlewareTest.php

jidbujawe committed:
Update ReviewTest.php

jidbujawe committed:
Passed requested tests and updated controller as necessary

jidbujawe committed:
Merge branch 'main' into jaydel/implement-reviews

# Conflicts:
#	tests/Http/Middleware/AdminMiddlewareTest.php

vrodriguero commented on app/Http/Requests/Review/UpdateReviewRequest.php:
```diff
@@ -0,0 +1,52 @@
+        } else {
+
+            return true;
+        }
```
I think that you do not need the else here Jayd ;D 

You can just have return true. Because if the condition is true this line will not be reached anyway :D

vrodriguero commented on app/Http/Requests/Review/CreateReviewRequest.php:
```diff
@@ -0,0 +1,39 @@
+
+use Illuminate\Foundation\Http\FormRequest;
+
+class CreateReviewRequest extends FormRequest
```
You can just call this Store

vrodriguero commented on app/Http/Requests/Review/UpdateReviewRequest.php:
```diff
@@ -0,0 +1,52 @@
+use Illuminate\Foundation\Http\FormRequest;
+use Illuminate\Http\Request;
+
+class UpdateReviewRequest extends FormRequest
```
You can call this just Update

vrodriguero commented on app/Http/Requests/ReviewRequest.php:
```diff
@@ -0,0 +1,49 @@
+            ]
+        ];
+    }
+}
```
Is this being used anywhere?
Otherwise you can delete this

vrodriguero commented on app/Http/Controllers/ReviewController.php:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    public function destroy(Review $review)
+    {
+        if (isset(auth()->user()->is_admin) && auth()->user()->is_admin)
+        {
+            return $this->allowDelete($review);
+        }
+
+        if (isset(auth()->user()->id) && auth()->user()->id == $review->user_id)
+        {
+            return $this->allowDelete($review);
+        }
+
+        return redirect()->route('reviews.show', $review)
+            ->with('status', 'Oops! You do not have permission to delete this review.');
+    }
```
maybe is easier to check the opposite?
if !auth()->user()->is_admin || auth()->user()->id === $review->user_id) then redirect, otherwise delete. 
That way you do not need an additional method for delete

jbujawe-cybercert replied:
Hey @vrodriguero, I'm not sure why but when I combine those two conditions together [auth()->user()->is_admin || auth()->user()->id === $review->user_id] without the isset() it does not work, it will restrict deletion even to the review owner, do you have any thoughts why that is Vick?

vrodriguero commented on routes/web.php:
```diff
@@ -1,10 +1,12 @@
 <?php
 
+
```
You can remove this extra line ;D

vrodriguero commented on tests/Http/Controllers/ReviewTest.php:
```diff
@@ -0,0 +1,424 @@
+    public function any_auth_user_cannot_edit_a_review_they_did_not_create()
+    {
+        $user = User::factory()->create([
+            'id' => 1
```
You can just use the create Jayd :D no need to specify the id here because we ate checking the id by $user->id

vrodriguero committed:
Merge branch 'main' into jaydel/implement-reviews

vrodriguero reviewed (changes requested):
just some suggestions Jayd! 
Awesome work

jidbujawe committed:
Implemented requested changes based on most recent PR review

jidbujawe committed:
Removed duplicated code from destroy method in Review controller

jidbujawe committed:
Merge branch 'main' into jaydel/implement-reviews

vrodriguero committed:
Merge branch 'jaydel/implement-reviews' of github.com:codinglabsau/moviedeck into jaydel/implement-reviews

jidbujawe committed:
Implemented requested changes

- used request()->user()->id
- used resource controller to change route path of /reviews to /movies/movie/reviews/review

jidbujawe committed:
Merge branch 'jaydel/implement-reviews' of https://github.com/codinglabsau/moviedeck into jaydel/implement-reviews

jidbujawe committed:
Merge branch 'main' into jaydel/implement-reviews

# Conflicts:
#	routes/web.php
#	tests/Http/Controllers/Auth/RegisterControllerTest.php
#	tests/Http/Controllers/CelebTest.php
#	tests/Http/Controllers/MovieTest.php

jidbujawe committed:
updated to pass celebs test

vrodriguero commented on resources/views/movies/show.blade.php:
```diff
@@ -97,20 +99,27 @@
                             <span class="font-normal text-sm"> {{ $review->created_at->diffForHumans() }} </span>
-                            <p class="py-2"> {{ $review->content }} </p>
+                            <p class="py-2">
+                                {{ \Illuminate\Support\Str::limit($review->content, 200, '...') }}
```
Perhaps this could be handle in the controller? Or maybe model if you use this in other places?
Maybe you can call this $review->excerpt?

vrodriguero replied:
So similar to what you have done with the movie duration 
You can add a getExcerptAttribute on the Review model :D and return the return in that format 😄

vrodriguero commented on tests/Http/Controllers/CelebTest.php:
```diff
@@ -181,6 +181,7 @@ public function guest_cannot_create_a_celeb()
     /** @test */
     public function admin_can_see_celebs_edit_view()
     {
+        $this->withoutExceptionHandling();
```
you can remove this ;D

vrodriguero reviewed (changes requested):
Amazing work Jayd, especially with the tricky tests ;D 

I have just a small suggestion for a refactor of the review content format used in some of the blade views :D.

Keep up with the great work!

jidbujawe committed:
Added review->excerpt

vrodriguero committed:
fix routes

vrodriguero committed:
remove duplicates

vrodriguero reviewed (approved):
Great Work Jayd!

I just removed a couple of duplicates cause because of the merge ;D
